File: Obsolete/CV_Test.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

int main() {
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Error: Cannot open camera." << std::endl;
        return -1;
    }

    cv::Mat frame, hsv, mask;

    cv::Scalar lower_hsv(0, 100, 100);
    cv::Scalar upper_hsv(10, 255, 255);

    while (true) {
        cap >> frame;
        if (frame.empty()) break;

        cv::cvtColor(frame, hsv, cv::COLOR_BGR2HSV);
        cv::inRange(hsv, lower_hsv, upper_hsv, mask);

        cv::erode(mask, mask, cv::Mat(), cv::Point(-1, -1), 2);
        cv::dilate(mask, mask, cv::Mat(), cv::Point(-1, -1), 2);

        std::vector<std::vector<cv::Point>> contours;
        cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

        for (const auto& contour : contours) {
            double area = cv::contourArea(contour);
            if (area > 500) {
                cv::Point2f center;
                float radius;
                cv::minEnclosingCircle(contour, center, radius);
                if (radius > 10) {
                    cv::circle(frame, center, (int)radius, cv::Scalar(0, 255, 0), 2);
                    cv::circle(frame, center, 3, cv::Scalar(0, 0, 255), -1);
                }
            }
        }

        cv::imshow("Ball Detection", frame);
        cv::imshow("Mask", mask);

        if (cv::waitKey(1) == 27) break;
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}

File: CV_Test.py
from picamera2 import Picamera2
import opencv2
import numpy as np

picam2 = Picamera2()
picam2.configure(picam2.preview_configuration(main={"format": "BGR888", "size": (640, 480)}))
picam2.start()

while True:
    frame = picam2.capture_array()

    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
    lower_orange = np.array([5, 150, 150])
    upper_orange = np.array([15, 255, 255])
    mask = cv2.inRange(hsv, lower_orange, upper_orange)

    contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    if contours:
        c = max(contours, key=cv2.contourArea)
        ((x, y), radius) = cv2.minEnclosingCircle(c)
        if radius > 10:
            cv2.circle(frame, (int(x), int(y)), int(radius), (0, 255, 0), 2)
            cv2.circle(frame, (int(x), int(y)), 5, (0, 0, 255), -1)
            print(f"Ball position: ({int(x)}, {int(y)})")

    cv2.imshow("Ball Tracking", frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

picam2.stop()
cv2.destroyAllWindows()
